api/src/opentrons/protocol_engine/state/pipettes.py

```diff
@@ -308,51 +309,42 @@ def _update_pipette_nozzle_map(
                 state_update.pipette_nozzle_map.pipette_id
             ] = state_update.pipette_nozzle_map.nozzle_map
 
-    def _update_volumes(
-        self, action: Union[SucceedCommandAction, FailCommandAction]
+    def _update_volumes(self, state_update: update_types.StateUpdate) -> None:
+        if state_update.pipette_aspirated_fluid == update_types.NO_CHANGE:
+            return
+        if state_update.pipette_aspirated_fluid.type == "aspirated":
+            self._update_aspirated(state_update.pipette_aspirated_fluid)
+        elif state_update.pipette_aspirated_fluid.type == "ejected":
+            self._update_ejected(state_update.pipette_aspirated_fluid)
+        elif state_update.pipette_aspirated_fluid.type == "empty":
+            self._update_empty(state_update.pipette_aspirated_fluid)
+        elif state_update.pipette_aspirated_fluid.type == "unknown":
+            self._update_unknown(state_update.pipette_aspirated_fluid)
+        else:
+            LOG.error(
+                f"Unknown aspirated fluid update type {state_update.pipette_aspirated_fluid.type}"
+            )
```

Comment:
We can turn this runtime problem into a static one:

```suggestion
            assert_never(state_update.pipette_aspirated_fluid)
```

With `assert_never` from `typing_extensions`.